File: 0240_Search_a_2D_Matrix_II/code4.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        i, j = 0, len(matrix[0]) - 1
        
        while i <= len(matrix) - 1 and j >= 0:
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] < target:
                i += 1
            else:
                j -= 1
                
        return False

File: 1466_Reorder_Routes_to_Make_All_Paths_Lead_to_the_City_Zero/code4.cpp
class Solution {
public:
    int minReorder(int n, vector<vector<int>>& connections) {
        
        // create graph
        unordered_map<int, vector<tuple<int, int>>> graph;
        for(int i = 0; i < connections.size(); i++){
            int node1 = connections[i][0], node2 = connections[i][1];
            graph[node2].push_back({node1, 0});
            graph[node1].push_back({node2, 1});
        }
        
        // traversal tree
        vector<int> visited(n, false);
        visited[0] = true;
        return recur(0, visited, graph);
    }
    
    int recur(int node, vector<int>& visited, unordered_map<int, vector<tuple<int, int>>>& graph){
        int ans = 0;
        for(auto& [nextNode, target]: graph[node]){
            if(!visited[nextNode]){
                visited[nextNode] = true;
                ans += target + recur(nextNode, visited, graph);
            }
        }
        return ans;
    }
};